### ISSUE TITLE
[FR] Add docstring for `from_pretrained`

### ISSUE THREAD

**Issue by PhilipMay** [huggingface/setfit#180]
https://github.com/huggingface/setfit/blob/99c30746799a09e0267427b8a7b8650568222b48/src/setfit/modeling.py#L311-L324

PS: I might provide a PR

**Comment by tomaarsen** [comment]
You are likely aware of this, but `_from_pretrained` is only used internally, and its docstring would never be exposed to users. 

Let me explain: `SetFitModel` subclasses [`PyTorchModelHubMixin`](https://github.com/huggingface/huggingface_hub/blob/31b5fde431bb8e87fbbf8ce1d35484c86a040a07/src/huggingface_hub/hub_mixin.py#L405) which subclasses [`ModelHubMixin`](https://github.com/huggingface/huggingface_hub/blob/31b5fde431bb8e87fbbf8ce1d35484c86a040a07/src/huggingface_hub/hub_mixin.py#L23) from `huggingface_hub`. Consequently, the docstring that users see under `SetFitModel.from_pretrained` is the one for [`ModelHubMixin.from_pretrained`](https://github.com/huggingface/huggingface_hub/blob/31b5fde431bb8e87fbbf8ce1d35484c86a040a07/src/huggingface_hub/hub_mixin.py#L101). Internally, this method calls `cls._from_pretrained`, i.e. the method from this repository that you have highlighted in this issue. 
So, updating the docstring of `SetFitModel._from_pretrained` does not change any docstring that users see.

 

I'm always in favour of docstrings, but I would argue that it has a low priority compared to other methods in `SetFitModel`, as it's not a method that users should interface with.

**Comment by PhilipMay** [comment]
@tomaarsen I agree. But from user perspective there should be some documentation about how to create the model.
To create it `from_pretrained` has to be used. So my question: How do you suggest to document that?

**Comment by tomaarsen** [comment]
I see two options that I would find satisfactory:

1. `SetFitModel.from_pretrained` is already documented via [`ModelHubMixin.from_pretrained`](https://github.com/huggingface/huggingface_hub/blob/31b5fde431bb8e87fbbf8ce1d35484c86a040a07/src/huggingface_hub/hub_mixin.py#L101). I believe that documentation is fine as is.
2. If we really want to update the documentation for `SetFitModel.from_pretrained`, then we must override or extend its docstring. It would not surprise me if some huggingface product already has functionality for this, but there's a handful of different approaches. Perhaps we can modify `SetFitModel.from_pretrained.__doc__` directly somehow, despite `__doc__` being write-locked, or we can simply create a method like so:
    ```python
        @classmethod
        def from_pretrained(cls, *args, **kwargs):
            """ New docstring here """
            return super().from_pretrained(*args, **kwargs)
    ```
    *However*, we then need to use some decorator tool like [functools.wraps](https://docs.python.org/3/library/functools.html#functools.wraps) or [wrapt](https://wrapt.readthedocs.io/en/latest/) to inherit the signature from the superclass method. In practice, we will also need to inherit the superclass method's docstring and add to it somehow, because the docstring should match the signature, and the signature may differ between different `huggingface_hub` versions. With other words, you would have to inherit the ["parameters" section](https://github.com/huggingface/huggingface_hub/blob/31b5fde431bb8e87fbbf8ce1d35484c86a040a07/src/huggingface_hub/hub_mixin.py#L115-L165), and you could update or replace the rest if you wanted. 

I have used functools.wraps and wrapt [some 3 years ago](https://github.com/tomaarsen/TheNounProjectAPI/blob/6db6031800047421146bc72e74060b8af9a7b3be/TheNounProjectAPI/call.py#L49), with the exact purpose of creating a decorator that inherits the signature. Updating or extending the docstring takes it one step further even.
In my experience, I found it to be way too much effort for what it was worth.

However, I've found a function that partially does what is needed for option 2, should we decide to go that route:
* [`replace_list_option_in_docstrings`](https://github.com/huggingface/transformers/blob/4eb918e656944df2757513c535e8ad8c01d632e2/src/transformers/models/auto/configuration_auto.py#L647-L668) definition.
* [Its use in practice](https://github.com/huggingface/transformers/blob/4eb918e656944df2757513c535e8ad8c01d632e2/src/transformers/models/auto/configuration_auto.py#L694-L705).

That said, I'm more than satisfied with option 1.

**Comment by tomaarsen** [comment]
I've changed my mind about this somewhat. I believe it is quite important to mention SetFit-specific parameters in `from_pretrained` in a docstring and such that an IDE can recognize e.g. `use_differentiable_head` as an actual parameter.

That said, I'm unsure what the best approach for this is.

- Tom Aarsen

**Comment by tomaarsen** [comment]
#439 should help. Although the IDE will not be much help, the upcoming documentation clearly states which arguments are allowed.

**Comment by tomaarsen** [comment]
Closed via #439
